File: app/src/main/java/com/friday/ar/plugin/installer/PluginInstaller.java
package com.friday.ar.plugin.installer;

import android.app.Notification;
import android.app.NotificationManager;
import android.content.Context;
import android.util.Log;

import androidx.core.app.NotificationCompat;

import com.friday.ar.Constant;
import com.friday.ar.R;
import com.friday.ar.plugin.file.Manifest;
import com.friday.ar.plugin.file.PluginFile;
import com.friday.ar.plugin.file.ZippedPluginFile;
import com.friday.ar.plugin.security.PluginVerifier;
import com.friday.ar.plugin.security.VerificationSecurityException;

import net.lingala.zip4j.exception.ZipException;

import org.json.JSONException;

import java.io.File;
import java.io.IOException;

public class PluginInstaller {
    private static final String LOGTAG = "PluginInstaller";
    private Context context;

    public PluginInstaller(Context context) {
        this.context = context;
    }

    public void installFrom(ZippedPluginFile pluginDir) throws IOException, IllegalFileException {
        File zippedPluginDefaultFile = pluginDir.getFile();
        try {
            pluginDir.extractAll(Constant.getPluginCacheDir(context).getPath());
            PluginFile extractionTargetDirectory = new PluginFile(Constant.getPluginCacheDir(context, zippedPluginDefaultFile.getName().replace(".fpl", "")).getPath());
            PluginVerifier.verify(extractionTargetDirectory, true);
            extractionTargetDirectory.renameTo(Constant.getPluginDir(context, zippedPluginDefaultFile.getName().replace(".fpl", "")));
        } catch (JSONException e) {
            Log.e(LOGTAG, e.getLocalizedMessage(), e);
            Notification notification = new NotificationCompat.Builder(context, Constant.NOTIF_CHANNEL_INSTALLER_ID)
                    .setContentTitle(context.getString(R.string.pluginInstaller_error_installation_failed))
                    .setContentText(context.getString(R.string.pluginInstaller_error_could_not_parse))
                    .setContentInfo(zippedPluginDefaultFile.getName())
                    .setSubText(context.getString(R.string.pluginInstaller_name))
                    .setCategory(NotificationCompat.CATEGORY_ERROR)
                    .setPriority(NotificationCompat.PRIORITY_HIGH)
                    .setSmallIcon(R.drawable.ic_twotone_error_outline_24px)
                    .setTicker(context.getString(R.string.pluginInstaller_error_could_not_parse))
                    .build();
            NotificationManager notificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
            notificationManager.notify(Constant.NotificationIDs.NOTIFICATION_INSTALL_ERROR, notification);
            return;
        } catch (VerificationSecurityException e) {
            Log.e(LOGTAG, e.getLocalizedMessage(), e);
            Notification notification = new NotificationCompat.Builder(context, Constant.NOTIF_CHANNEL_INSTALLER_ID)
                    .setContentTitle(context.getString(R.string.pluginInstaller_error_installation_failed))
                    .setContentText(context.getString(R.string.pluginInstaller_error_manifest_security))
                    .setContentInfo(zippedPluginDefaultFile.getName())
                    .setSubText(context.getString(R.string.pluginInstaller_name))
                    .setCategory(NotificationCompat.CATEGORY_ERROR)
                    .setPriority(NotificationCompat.PRIORITY_HIGH)
                    .setSmallIcon(R.drawable.ic_twotone_error_outline_24px)
                    .setTicker(context.getString(R.string.pluginInstaller_error_could_not_parse))
                    .build();
            NotificationManager notificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
            notificationManager.notify(Constant.NotificationIDs.NOTIFICATION_INSTALL_ERROR, notification);
            return;
        } catch (ZipException e) {
            Log.e(LOGTAG, e.getLocalizedMessage(), e);
        }
        new File(zippedPluginDefaultFile.getPath()).renameTo(new File(context.getFilesDir() + "/plugin/" + zippedPluginDefaultFile.getName()));
        try {
            Manifest succesPluginManifest = new PluginFile(Constant.getPluginDir(context, zippedPluginDefaultFile.getName().replace(".fpl", "")).getPath()).getManifest();
            Notification notification = new NotificationCompat.Builder(context, Constant.NOTIF_CHANNEL_INSTALLER_ID)
                    .setContentTitle(context.getString(R.string.pluginInstaller_succes_install_title))
                    .setContentText(zippedPluginDefaultFile.getName())
                    .setSubText(context.getString(R.string.pluginInstaller_name))
                    .setCategory(NotificationCompat.CATEGORY_STATUS)
                    .setPriority(NotificationCompat.PRIORITY_DEFAULT)
                    .setSmallIcon(R.drawable.ic_twotone_save_alt_24px)
                    .setTicker(context.getString(R.string.pluginInstaller_success_ticker_text, succesPluginManifest.getPluginName()))
                    .build();
            NotificationManager notificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
            notificationManager.notify(Constant.NotificationIDs.NOTIFICATION_INSTALL_SUCCESS, notification);
        } catch (JSONException e) {
            Log.e(LOGTAG, e.getLocalizedMessage(), e);
        }
    }

    public class IllegalFileException extends IllegalStateException {
        IllegalFileException(String message) {
            super(message);
        }
    }
}

File: app/src/main/java/com/friday/ar/plugin/installer/PluginInstaller.kt
package com.friday.ar.plugin.installer

import android.content.Context
import android.util.Log
import com.friday.ar.Constant
import com.friday.ar.FridayApplication
import com.friday.ar.R
import com.friday.ar.plugin.Plugin
import com.friday.ar.plugin.file.ZippedPluginFile
import com.friday.ar.plugin.security.PluginVerifier
import com.friday.ar.plugin.security.VerificationSecurityException
import com.friday.ar.util.FileUtil
import com.friday.ar.util.cache.PluginFileCacheUtil
import net.lingala.zip4j.exception.ZipException
import org.json.JSONException
import java.io.IOException
import java.nio.file.Files
import java.nio.file.StandardCopyOption

class PluginInstaller(private val context: Context) {
    companion object {
        private const val LOGTAG = "PluginInstaller"
    }

    private var onInstallationStateChangedListenerList: ArrayList<OnInstallationStateChangedListener> = ArrayList()

    @Throws(IOException::class)
    fun installFrom(pluginDir: ZippedPluginFile) {
        val verifier = PluginVerifier()
        val cachedPluginFile = PluginFileCacheUtil.cachePluginFile(context, pluginDir)
        verifier.setOnVerificationCompleteListener(object : PluginVerifier.OnVerificationCompleteListener {
            override fun onSuccess() {
                notifyInstallProgressChange(context.getString(R.string.pluginInstaller_progressMessage_installing))
                Files.move(cachedPluginFile.toPath(), Constant.getPluginDir(context, cachedPluginFile.name).toPath(), StandardCopyOption.REPLACE_EXISTING)
                notifyInstallSuccess()
                cachedPluginFile.deleteRecursively()
            }

            override fun onZipException(e: ZipException) {
                Log.e(LOGTAG, e.message, e)
                notifyInstallationFailed(e)
            }

            override fun onIOException(e: IOException) {
                Log.e(LOGTAG, e.message, e)
                notifyInstallationFailed(e)
            }

            override fun onJSONException(e: JSONException) {
                Log.e(LOGTAG, e.message, e)
                notifyInstallationFailed(e)
            }

            override fun onVerificationFailed(e: VerificationSecurityException) {
                Log.e(LOGTAG, e.message, e)
                notifyInstallationFailed(e)
            }
        })
        notifyInstallProgressChange(context.getString(R.string.pluginInstaller_progressMessage_verifying))
        verifier.verify(cachedPluginFile, true)
    }

    fun uninstallPlugin(plugin: Plugin) {
        Log.i(LOGTAG, "uninstalling plugin:${plugin.name}")
        FileUtil.deleteDirectory(plugin.pluginFile)
        (context.applicationContext as FridayApplication).applicationPluginLoader!!.indexedPlugins.remove(plugin)
    }

    fun addOnInstallationProgressChangedListener(onInstallationStateChangedListener: OnInstallationStateChangedListener) {
        onInstallationStateChangedListenerList.add(onInstallationStateChangedListener)
    }

    fun removeOnInstallationStateChangedListener(onInstallationStateChangedListener: OnInstallationStateChangedListener) {
        onInstallationStateChangedListenerList.remove(onInstallationStateChangedListener)
    }

    private fun notifyInstallProgressChange(msg: String) {
        onInstallationStateChangedListenerList.forEach { listener ->
            listener.onProgressChanged(msg)
        }
    }

    private fun notifyInstallSuccess() {
        onInstallationStateChangedListenerList.forEach { listener ->
            listener.onSuccess()
        }
    }

    private fun notifyInstallationFailed(e: Exception) {
        onInstallationStateChangedListenerList.forEach { listener ->
            listener.onFailure(e)
        }
    }

    interface OnInstallationStateChangedListener {
        fun onProgressChanged(progressMessage: String)
        fun onSuccess()
        fun onFailure(e: Exception)
    }
}
